modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
       logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
       throw new IndexRebuildException(index.getIndexName(), getService(), e);
     }
-    final int limit = 1000;
 
     if (total > 0) {
       logIndexRebuildBegin(logger.getSlf4jLogger(), index.getIndexName(), total, "workflows");
+      List<WorkflowIndexData> workflowIndexData;
+
+      try {
+        workflowIndexData = persistence.getWorkflowIndexData();
+      } catch (WorkflowDatabaseException e) {
+        logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
+        throw new IndexRebuildException(index.getIndexName(), getService(), e);
+      }
+
+      Map<String, WorkflowIndexData> latestWorkflows = new HashMap<>();
+      for (WorkflowIndexData data : workflowIndexData) {
+        latestWorkflows.put(data.getMediaPackageId(), data);
+      }
+
       int current = 0;
-      int offset = 0;
-      List<WorkflowInstance> workflows;
       do {
-        try {
-          workflows = persistence.getWorkflowInstances(limit, offset);
-        } catch (Exception e) {
-          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), total, current, e);
-          throw new IndexRebuildException(index.getIndexName(), getService(), e);
-        }
-        logger.debug("Got {} workflows for re-indexing", workflows.size());
-        offset += limit;
-
-        for (WorkflowInstance instance : workflows) {
+        logger.debug("Got {} workflows for re-indexing", latestWorkflows.size());
+        for (WorkflowIndexData data : latestWorkflows.values()) {
           current += 1;
           Organization organization = null;
           try {
-            organization = organizationDirectoryService.getOrganization(instance.getOrganizationId());
+            organization = organizationDirectoryService.getOrganization(data.getOrganizationId());
           } catch (NotFoundException e) {
-            logger.error("Found workflow with non-existing organization {}", instance.getOrganizationId());
+            logger.error("Found workflow with non-existing organization {}", data.getOrganizationId());
             continue;
           }
 
-          // get metadata for index update
-          final DublinCoreCatalog episodeDublinCoreCatalog = getEpisodeDublinCoreCatalog(instance.getMediaPackage());
-
-          // get acl for active workflows.
-          // don't try this for terminated workflows since the ACLs are no longer in the working file repository and
-          // they will be overwritten later in the re-indexing process by ACLs from the asset manager anyway.
-          final AccessControlList accessControlList;
-          if (instance.getState().isTerminated()) {
-            accessControlList = new AccessControlList();
-          } else {
-            accessControlList = authorizationService.getActiveAcl(instance.getMediaPackage()).getA();
-          }
-
           SecurityUtil.runAs(securityService, organization,
```

Comment:
If we don't pass modify the creator, I don't think we need to change the security context.